src/raft.rs

```diff
@@ -884,16 +994,132 @@ impl<T: Storage> Raft<T> {
         self.r.send_append_aggressively(to, pr, &mut self.msgs)
     }
 
+    // Find an appropriate agent.
+    // If found an appropriate one, mark the corresponding message's type as
+    // MsgGroupBroadcast and return true. If not found, return false.
+    fn select_agent_for_bcast_group(&self, msgs: &mut [Message]) -> bool {
+        for msg in msgs {
+            let peer_id = msg.to;
+            let is_voter = self.prs().conf().voters().contains(peer_id);
+            // Agent must be voter and recently active.
+            if !is_voter || !self.is_recent_active(peer_id) {
```

Comment:
These information can be obtained when the message is generated.

Reply:
The information of `voter` and `recent_active` is only used for agent selection. When the message is generated, we only determine whether it should be replicated by follower.
If we want to obtain these information when the message is generated, another cache is needed to record them.

Reply:
No, what you need to do is merging `bcast_append`, `merge_msg_group` and `select_agent_for_bcast_group`.